models/llama3/reference_impl/generation.py

```diff
@@ -134,25 +141,29 @@ def build(
         )
         tokenizer = Tokenizer(model_path=tokenizer_path)
         assert model_args.vocab_size == tokenizer.n_words
-        if torch.cuda.is_bf16_supported():
+        if _is_cuda(device) and torch.cuda.is_bf16_supported():
             torch.set_default_tensor_type(torch.cuda.BFloat16Tensor)
-        else:
+        elif _is_cuda(device):
             torch.set_default_tensor_type(torch.cuda.HalfTensor)
+        else:
+            torch.set_default_tensor_type(torch.float16)
+        
         if model_args.vision_chunk_size > 0:
             from .multimodal.model import CrossAttentionTransformer
 
-            model = CrossAttentionTransformer(model_args)
+            model = CrossAttentionTransformer(model_args, device=device)
             model.setup_cache(model_args.max_batch_size, torch.bfloat16)
         else:
-            model = Transformer(model_args)
+            model = Transformer(model_args, device=device)
```

Comment:
On my side that's not enough to apply this PR and pass `device` to `Transforner` on initialization. I still need to call `model.to(device)` in a consequent step. I.e.:
```
model = Transformer(model_args, device=device)
model.to(device)
```

This does not quite makes sense to me though. If we pass `device` on class creation then we should not require to still call `.to()` to cast everything to the same device. I guess it implies that either `.to()` should be done in the end of `__init__()` of `Transformer` or passing `device` should completely be not required and initialization of classes fixed in a way that they are device agnostic (created on CPU), then `.to()` should just work since it seem to run recursively for all submodules. I think the reason it did not work is couple places where `.cuda()` is called on tensor creation.

Reply:
Hm, it does work on my machine (though I recognize that's not especially helpful). Specifically, I ran this script with and without this line commented (the `.to()` call) and saw identical results:

https://github.com/anordin95/run-llama-locally/blob/198613041e56e28cd9262cf30b5365a229e56874/minimal_run_inference.py#L35

Can you share the error message you're encountering?

Reply:
I met this behavior trying to enable xpu pytorch backend on llama_stack. Steps I've made are described at https://github.com/dvrogozh/notebook/blob/master/meta-llama/run-llama-stack-on-intel-gpu.md. It's required to apply your patch for llama-models and another patch for llama-stack. The latter patch (for llama-stack) I've implemented for XPU device, but it should be fairly trivial to try adding mps.

The relevant line with the `.to()` call is here:
https://github.com/meta-llama/llama-stack/blob/cc788054276114390871e5172b1b1e360f14b365/llama_stack/providers/inline/inference/meta_reference/generation.py#L200

```
Traceback (most recent call last):
  File "/home/dvrogozh/git/llama-stack/llama_stack/distribution/server/server.py", line 187, in endpoint
    return await maybe_await(value)
  File "/home/dvrogozh/git/llama-stack/llama_stack/distribution/server/server.py", line 151, in maybe_await
    return await value
  File "/home/dvrogozh/git/llama-stack/llama_stack/distribution/routers/routers.py", line 123, in chat_completion
    return await provider.chat_completion(**params)
  File "/home/dvrogozh/git/llama-stack/llama_stack/providers/inline/inference/meta_reference/inference.py", line 234, in chat_completion
    return await self._nonstream_chat_completion(request)
  File "/home/dvrogozh/git/llama-stack/llama_stack/providers/inline/inference/meta_reference/inference.py", line 276, in _nonstream_chat_completion
    return impl()
  File "/home/dvrogozh/git/llama-stack/llama_stack/providers/inline/inference/meta_reference/inference.py", line 244, in impl
    for token_result in self.generator.chat_completion(request):
  File "/home/dvrogozh/git/llama-stack/llama_stack/providers/inline/inference/meta_reference/model_parallel.py", line 93, in chat_completion
    yield from gen
  File "/home/dvrogozh/git/llama-stack/llama_stack/providers/inline/inference/meta_reference/parallel_utils.py", line 368, in run_inference
    raise Exception(obj.error)
Exception: Expected all tensors to be on the same device, but found at least two devices, cpu and xpu:0! (when checking argument for argument index in method wrapper_XPU__index_select)
```

Reply:
@anordin95 : I tried your [script](https://github.com/anordin95/run-llama-locally/blob/198613041e56e28cd9262cf30b5365a229e56874/minimal_run_inference.py#L35) and I actually see the same error on XPU backend (if `.to()` is not called) as I shared above with llama-stack:
```
> initializing model parallel with size 1
> initializing ddp with size 1
> initializing pipeline with size 1
[rank0]: Traceback (most recent call last):
[rank0]:   File "/home/dvrogozh/tmp/llm.py", line 48, in <module>
[rank0]:     output_batch = llama_model(input_batch, start_pos=0)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/home/dvrogozh/git/llama-models/models/llama3/reference_impl/model.py", line 306, in forward
[rank0]:     h = self.tok_embeddings(tokens)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/fairscale/nn/model_parallel/layers.py", line 136, in forward
[rank0]:     output_parallel = F.embedding(
[rank0]:   File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/nn/functional.py", line 2551, in embedding
[rank0]:     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
[rank0]: RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and xpu:0! (when checking argument for argument index in method wrapper_XPU__index_select)
```

Reply:
@anordin95 : you've tried with `mps` backend, right? (by default script runs with `cpu` which indeed works with and without `.to()`)

Reply:
@dvrogozh  Ah, I had not. Silly me. I get the same error as you when trying `mps`. I think there a few problems. 

One is the `tok_embeddings`:
https://github.com/anordin95/llama-models/blob/13563f301f034a418aa6a9806f05c29edf1e6715/models/llama3/reference_impl/model.py#L283-L285
These rely on a separate library and the API currently does not expose a device option.

Another is the RMSNorm:
https://github.com/anordin95/llama-models/blob/13563f301f034a418aa6a9806f05c29edf1e6715/models/llama3/reference_impl/model.py#L31-L35
I modified the `RMSNorm__init__` to pass a `device` parameter, however an earlier step, loading the saved model-weights, then failed. 

Either way, I'm planning to close this PR since I don't think the Meta team wants these changes.